### PR TITLE
refactor array_analysis

### PR DESCRIPTION
**Pull request by sklam**
Refactoring to make  array_analysis.py easier to follow.

Main changes:

* `ArrayAnalysis.run()` is divided up into two functions. The first function is gather all the transformation. The second function applies the transformation. The first function still mutates side-states like calltypes and typemap.
* `class Signature` in typing module is now immutable.
* Lint array_analysis.py with `black`.

### CODE REVIEW

File: numba/parfors/array_analysis.py [hunk 1]
```diff
@@ -113,15 +134,22 @@ def assert_equiv(typingctx, *val):
...
+    assert all(
+        map(
+            lambda a: (
+                isinstance(a, types.ArrayCompatible)
+                or isinstance(a, types.BaseTuple)
+                or isinstance(a, types.SliceType)
+                or isinstance(a, types.Integer)
+            ),
+            val[0][1:],
+        )
```
Comment (eric-wieser) [inline numba/parfors/array_analysis.py]: nit:
```suggestion
        isinstance(a, types.ArrayCompatible)
        or isinstance(a, types.BaseTuple)
        or isinstance(a, types.SliceType)
        or isinstance(a, types.Integer)
        for a in val[0][1:]
```

File: numba/parfors/array_analysis.py [hunk 2]
```diff
@@ -367,10 +409,14 @@ def _get_names(self, obj):
                 return (name,)
 
             typ = self.typemap[name]
-            if (isinstance(typ, types.BaseTuple) or
-                    isinstance(typ, types.ArrayCompatible)):
-                ndim = (typ.ndim if isinstance(typ, types.ArrayCompatible)
-                        else len(typ))
+            if isinstance(typ, types.BaseTuple) or isinstance(
+                typ, types.ArrayCompatible
+            ):
```
Comment (eric-wieser) [inline numba/parfors/array_analysis.py]:
```suggestion
            if isinstance(typ, (types.BaseTuple, types.ArrayCompatible)):
```

File: numba/parfors/array_analysis.py [hunk 3]
```diff
@@ -1009,90 +1105,125 @@ def run(self, blocks=None, equiv_set=None):
...
 
         cfg = compute_cfg_from_blocks(blocks)
         topo_order = find_topo_order(blocks, cfg=cfg)
         # Traverse blocks in topological order
+        self._run_on_blocks(topo_order, blocks, cfg, init_equiv_set)
+
+        if config.DEBUG_ARRAY_OPT >= 1:
+            self.dump()
+            print(
+                "ArrayAnalysis post variable types: ",
```
Comment (stuartarchibald) [inline numba/parfors/array_analysis.py]: What's the "post" referring to and how is it different to https://github.com/numba/numba/pull/5320/files#diff-a46f9b8e7e173f928f6f58a2b9af202cR1109 ? (I know the answer, just wonder if we can make it clearer?).

File: numba/parfors/array_analysis.py [hunk 4]
```diff
@@ -1177,43 +1323,49 @@ def _analyze_inst(self, label, scope, equiv_set, inst, redefined):
...
+                self._index_to_shape, scope, equiv_set, inst.target, index
+            )
             if not result:
                 return [], []
             if result[0] is not None:
-                assert(isinstance(inst, ir.SetItem))
+                assert isinstance(inst, ir.SetItem)
                 inst.index = result[0]
-                #inst.index_var = result[0]
+                # inst.index_var = result[0]
```
Comment (stuartarchibald) [inline numba/parfors/array_analysis.py]: delete?

File: numba/parfors/array_analysis.py [hunk 5]
```diff
@@ -1315,53 +1481,95 @@ def _analyze_op_exhaust_iter(self, scope, equiv_set, expr):
...
-        the_var, the_typ = self.gen_literal_slice_part(rhs_rel - lhs_rel,
-                                                       loc,
-                                                       scope,
-                                                       stmts,
-                                                       equiv_set,
-                                                       name="static_slice_size")
+    def gen_static_slice_size(
+        self, lhs_rel, rhs_rel, loc, scope, stmts, equiv_set
+    ):
+        the_var, the_typ = self.gen_literal_slice_part(
```
Comment (stuartarchibald) [inline numba/parfors/array_analysis.py]: `the_var, *_ =` perhaps? `the_typ` isn't used?

File: numba/parfors/array_analysis.py [hunk 6]
```diff
@@ -1439,9 +1673,25 @@ def slice_size(self, index, dsize, equiv_set, scope, stmts):
...
+                "lhs=",
+                lhs,
+                "rhs=",
+                rhs,
+                "size_typ=",
+                size_typ,
+                "lhs_typ=",
+                lhs_typ,
+                "rhs_typ=",
+                rhs_typ,
```
Comment (stuartarchibald) [inline numba/parfors/array_analysis.py]: This is particularly weird format wise.

Comment (sklam) [reply numba/parfors/array_analysis.py]: oh boy, this is weird

**Pull request merged**